libraries/anvill_passes/src/InstructionFolderPass.cpp

```diff
@@ -0,0 +1,632 @@
+  bool function_changed{false};
+
+  while (!next_worklist.empty()) {
+    worklist.swap(next_worklist);
+    next_worklist.clear();
+
+    for (auto instr : worklist) {
+      // Keep track of the instructions we have visited so we don't
+      // loop forever
+      if (visited_instructions.count(instr) != 0U) {
+        continue;
+      }
+
+      visited_instructions.insert(instr);
+
+      // Attempt to combine this instruction; the folder function also
+      // drops all the instructions that are no longer needed but will
+      // keep `instr` alive for now
+      auto instr_folder_it = kInstructionFolderMap.find(instr->getOpcode());
+      if (instr_folder_it != kInstructionFolderMap.end()) {
+        const auto &instr_folder = instr_folder_it->second;
+        if (instr_folder(next_worklist, instr)) {
+          function_changed = true;
+        }
+      }
+    }
+  }
+
+  // Go through the visited instructions and drop the ones we no
+  // longer need. Doing the cleanup now ensures we do not risk
+  // new instructions polluting the unordered_set by ending up
+  // allocated at a memory location we have already seen
+  for (auto &instr : visited_instructions) {
+    if (instr->use_empty()) {
+      instr->eraseFromParent();
+    }
+  }
+
+  return function_changed;
+}
+
+llvm::StringRef InstructionFolderPass::getPassName(void) const {
+  return llvm::StringRef("InstructionFolderPass");
+}
+
+bool InstructionFolderPass::FoldSelectInstruction(
+    InstructionFolderPass::InstructionList &output, llvm::Instruction *instr) {
+
+  // Extract the condition and the two operands for later
+  llvm::Value *condition{nullptr};
+  llvm::Value *true_value{nullptr};
+  llvm::Value *false_value{nullptr};
+
+  {
+    auto select_instr = llvm::dyn_cast<llvm::SelectInst>(instr);
+
+    condition = select_instr->getCondition();
+    true_value = select_instr->getTrueValue();
+    false_value = select_instr->getFalseValue();
+  }
+
+  // Since we are iterating over the users, we have to postpone
+  // the deletion after the loop
+  InstructionList replaced_instr_list;
+
+  // Go through all the users of this `select` instruction
+  for (auto user : instr->users()) {
+    auto user_instr = llvm::dyn_cast<llvm::Instruction>(user);
+
+    // Search for a function that knows how to handle this case
+    auto instr_folder_it =
+        kSelectInstructionFolderMap.find(user_instr->getOpcode());
+
+    if (instr_folder_it == kSelectInstructionFolderMap.end()) {
+      continue;
+    }
+
+    const auto &instr_folder = instr_folder_it->second;
+
+    // If we succeed in folding this instruction, then also perform
+    // the replacement and the required cleanup
+    llvm::Instruction *replacement{nullptr};
+    if (!instr_folder(replacement, instr, condition, true_value, false_value,
+                      user_instr)) {
+      continue;
+    }
+
+    user_instr->replaceAllUsesWith(replacement);
+
+    replaced_instr_list.push_back(user_instr);
```

Comment:
This is redundant; `replaceAllUsesWith` means there are no uses left, so you can do `replaced_instr->eraseFromParent();` here without the later loop and `use_empty` check.

Reply:
If you want to be super safe (maybe there is some weird PHI node issue) then what I'd recommend is that `replace_instr_list` be a list a `std::pair`, and you `emplace_back(user_instr, replacement)`, then in your later loop, you do:

```c++
for (auto [instr, replacement_instr] : replaced_instr_list) {
  instr->replaceAllUsesWith(replacement_instr);
  instr->eraseFromParent();
}
```